File: frontend/src/lib/player/moov.ts
/**
 * Parse the moov atom from an MP4 file header to extract video FPS.
 * Fetches only the first 2 MB of the video via HTTP Range request.
 *
 * Navigates: moov → trak (video, with vmhd or 'vide' hdlr in mdia/minf)
 *   → mdia → mdhd (timescale) + stts (sample_delta)
 * FPS = timescale / sample_delta
 */

interface AtomRange {
  start: number;
  end: number; // exclusive
}

function readU32BE(data: Uint8Array, offset: number): number {
  return (
    (data[offset] << 24) |
    (data[offset + 1] << 16) |
    (data[offset + 2] << 8) |
    data[offset + 3]
  ) >>> 0;
}

function readU64BE(data: Uint8Array, offset: number): bigint {
  const hi = readU32BE(data, offset);
  const lo = readU32BE(data, offset + 4);
  return (BigInt(hi) << 32n) | BigInt(lo);
}

function atomSize(data: Uint8Array, offset: number): { size: bigint; headerLen: number } {
  const s = BigInt(readU32BE(data, offset));
  if (s === 1n) {
    return { size: readU64BE(data, offset + 8), headerLen: 16 };
  } else if (s === 0n) {
    return { size: BigInt(data.length - offset), headerLen: 8 };
  } else {
    return { size: s, headerLen: 8 };
  }
}

function fourccAt(data: Uint8Array, offset: number): string {
  return String.fromCharCode(
    data[offset],
    data[offset + 1],
    data[offset + 2],
    data[offset + 3],
  );
}

function findAtom(data: Uint8Array, start: number, fourcc: string): AtomRange | null {
  let offset = start;
  while (offset + 8 <= data.length) {
    const { size, headerLen } = atomSize(data, offset);
    if (size < BigInt(headerLen)) break;
    const atomType = fourccAt(data, offset + 4);
    const end = offset + Number(size);
    if (atomType === fourcc) {
      return { start: offset, end: Math.min(end, data.length) };
    }
    if (end > data.length) break;
    offset = end;
  }
  return null;
}

function findChildAtom(
  data: Uint8Array,
  parent: AtomRange,
  fourcc: string,
): AtomRange | null {
  const headerLen = readU32BE(data, parent.start) === 1 ? 16 : 8;
  return findAtom(data, parent.start + headerLen, fourcc);
}

/// Iterate all direct children of `parent` matching `fourcc`.
function* iterChildAtoms(
  data: Uint8Array,
  parent: AtomRange,
  fourcc: string,
): Generator<AtomRange> {
  const headerLen = readU32BE(data, parent.start) === 1 ? 16 : 8;
  let offset = parent.start + headerLen;
  const limit = Math.min(parent.end, data.length);
  while (offset + 8 <= limit) {
    const { size } = atomSize(data, offset);
    if (size < 8n) break;
    const atomType = fourccAt(data, offset + 4);
    const end = offset + Number(size);
    const clampedEnd = Math.min(end, limit);
    if (atomType === fourcc) {
      yield { start: offset, end: clampedEnd };
    }
    if (end > limit) break;
    offset = end;
  }
}

/// Check if a trak is a video track.
/// A video track has either `vmhd` in `minf`, or `hdlr` with handler type `vide`.
function isVideoTrak(data: Uint8Array, trak: AtomRange): boolean {
  const mdia = findChildAtom(data, trak, 'mdia');
  if (!mdia) return false;

  // Check hdlr for 'vide' (handler type at offset 20 = 8B header + 1B ver + 3B flags + 4B type + 4B subtype)
  const hdlr = findChildAtom(data, mdia, 'hdlr');
  if (hdlr) {
    const handlerType = fourccAt(data, hdlr.start + 20);
    if (handlerType === 'vide') return true;
  }

  // Check minf for vmhd
  const minf = findChildAtom(data, mdia, 'minf');
  if (minf && findChildAtom(data, minf, 'vmhd')) return true;

  return false;
}

export async function extractFpsFromUrl(videoSrc: string): Promise<number | null> {
  try {
    const resp = await fetch(videoSrc, {
      headers: { Range: 'bytes=0-2097151' }, // first 2 MB
    });
    if (!resp.ok) {
      console.warn('[moov] fetch failed:', resp.status, resp.statusText);
      return null;
    }

    const buf = await resp.arrayBuffer();
    const data = new Uint8Array(buf);
    console.log('[moov] fetched', data.length, 'bytes, first fourcc:', fourccAt(data, 4));

    const moov = findAtom(data, 0, 'moov');
    if (!moov) {
      console.warn('[moov] moov atom not found in first', data.length, 'bytes');
      return null;
    }
    console.log('[moov] found moov at', moov.start, 'size', moov.end - moov.start);

    // Find video trak
    let videoTrak: AtomRange | null = null;
    for (const trak of iterChildAtoms(data, moov, 'trak')) {
      if (isVideoTrak(data, trak)) {
        videoTrak = trak;
        break;
      }
    }
    if (!videoTrak) {
      console.warn('[moov] no video trak found');
      return null;
    }

    const mdia = findChildAtom(data, videoTrak, 'mdia');
    if (!mdia) { console.warn('[moov] mdia not found'); return null; }

    const mdhd = findChildAtom(data, mdia, 'mdhd');
    if (!mdhd) { console.warn('[moov] mdhd not found'); return null; }

    const timescale = readU32BE(data, mdhd.start + 20);

    const minf = findChildAtom(data, mdia, 'minf');
    if (!minf) { console.warn('[moov] minf not found'); return null; }
    const stbl = findChildAtom(data, minf, 'stbl');
    if (!stbl) { console.warn('[moov] stbl not found'); return null; }
    const stts = findChildAtom(data, stbl, 'stts');
    if (!stts) { console.warn('[moov] stts not found'); return null; }

    // stts entries: [4B sample_count][4B sample_delta]...
    // first sample_delta is at byte 20 (8B header + 1B ver + 3B flags + 4B count + 4B sample_count)
    const sampleDelta = readU32BE(data, stts.start + 20);
    console.log('[moov] timescale:', timescale, 'sampleDelta:', sampleDelta);

    if (sampleDelta === 0 || timescale === 0) {
      console.warn('[moov] zero timescale or sampleDelta');
      return null;
    }

    const fps = Math.round(timescale / sampleDelta);
    console.log('[moov] detected fps:', fps);
    return fps;
  } catch (e) {
    console.warn('[moov] exception:', e);
    return null;
  }
}

File: backend/src/services/moov.rs
use anyhow::{anyhow, Result};

pub struct MoovInfo {
    pub fps: f32,
}

/// Parse the first bytes of an MP4 file to extract FPS from the moov atom.
/// Expects the moov atom to be at the beginning of the data (faststart / web-optimized).
/// Finds the VIDEO track (with vmhd) and extracts: mdhd.timescale / stts.sample_delta.
pub fn parse_fps(data: &[u8]) -> Result<MoovInfo> {
    let moov = find_atom(data, 0, b"moov")
        .ok_or_else(|| anyhow!("moov atom not found — video may not be faststart"))?;

    let trak = find_video_trak(data, moov)
        .ok_or_else(|| anyhow!("video trak (with vmhd) not found in moov"))?;

    let mdia = find_child_atom(data, trak, b"mdia")
        .ok_or_else(|| anyhow!("mdia atom not found in trak"))?;

    let mdhd_range = find_child_atom(data, mdia, b"mdhd")
        .ok_or_else(|| anyhow!("mdhd atom not found in mdia"))?;

    // mdhd box: timescale at byte 20 (8B header + 1B ver + 3B flags + 4B+4B times)
    let timescale = read_u32_be(data, mdhd_range.0 + 20);

    let minf = find_child_atom(data, mdia, b"minf")
        .ok_or_else(|| anyhow!("minf atom not found in mdia"))?;
    let stbl = find_child_atom(data, minf, b"stbl")
        .ok_or_else(|| anyhow!("stbl atom not found in minf"))?;
    let stts_range = find_child_atom(data, stbl, b"stts")
        .ok_or_else(|| anyhow!("stts atom not found in stbl"))?;

    // stts entries: [4B sample_count][4B sample_delta]...
    // first sample_delta at byte 20 (8B header + 1B ver + 3B flags + 4B count + 4B sample_count)
    let sample_delta = read_u32_be(data, stts_range.0 + 20);

    if sample_delta == 0 {
        return Err(anyhow!("stts sample_delta is zero — cannot compute fps"));
    }

    let fps = (timescale as f32 / sample_delta as f32).round();

    Ok(MoovInfo { fps })
}

// ── low-level helpers ──────────────────────────────────────────────────────────

type AtomRange = (usize, usize); // (start, end) — end is exclusive

fn read_u32_be(data: &[u8], offset: usize) -> u32 {
    u32::from_be_bytes([
        data[offset],
        data[offset + 1],
        data[offset + 2],
        data[offset + 3],
    ])
}

fn read_u64_be(data: &[u8], offset: usize) -> u64 {
    u64::from_be_bytes([
        data[offset],
        data[offset + 1],
        data[offset + 2],
        data[offset + 3],
        data[offset + 4],
        data[offset + 5],
        data[offset + 6],
        data[offset + 7],
    ])
}

fn atom_size(data: &[u8], offset: usize) -> (u64, usize) {
    let s = read_u32_be(data, offset) as u64;
    if s == 1 {
        (read_u64_be(data, offset + 8), 16)
    } else if s == 0 {
        ((data.len() - offset) as u64, 8)
    } else {
        (s, 8)
    }
}

fn find_atom(data: &[u8], start: usize, fourcc: &[u8; 4]) -> Option<AtomRange> {
    let mut offset = start;
    while offset + 8 <= data.len() {
        let (size, header_len) = atom_size(data, offset);
        if size < header_len as u64 {
            break;
        }
        let atom_type: &[u8; 4] = data[offset + 4..offset + 8].try_into().unwrap();
        let end = offset + size as usize;
        if atom_type == fourcc {
            return Some((offset, end.min(data.len())));
        }
        if end > data.len() {
            break;
        }
        offset = end;
    }
    None
}

fn find_child_atom(data: &[u8], parent: AtomRange, fourcc: &[u8; 4]) -> Option<AtomRange> {
    let header_len = if read_u32_be(data, parent.0) == 1 { 16 } else { 8 };
    find_atom(data, parent.0 + header_len, fourcc)
}

/// Check if a trak is a video track (has `vmhd` in minf or `hdlr` type `vide`).
fn is_video_trak(data: &[u8], trak: AtomRange) -> bool {
    if let Some(mdia) = find_child_atom(data, trak, b"mdia") {
        // Check hdlr for 'vide'
        if let Some(hdlr) = find_child_atom(data, mdia, b"hdlr") {
            // handler type is at byte 20 of hdlr
            if hdlr.0 + 20 + 4 <= data.len() {
                let htype: &[u8; 4] = data[hdlr.0 + 20..hdlr.0 + 24].try_into().unwrap();
                if htype == b"vide" {
                    return true;
                }
            }
        }
        // Check minf for vmhd
        if let Some(minf) = find_child_atom(data, mdia, b"minf") {
            if find_child_atom(data, minf, b"vmhd").is_some() {
                return true;
            }
        }
    }
    false
}

/// Find the video `trak` — the one that passes `is_video_trak`.
fn find_video_trak(data: &[u8], moov: AtomRange) -> Option<AtomRange> {
    let header_len = if read_u32_be(data, moov.0) == 1 { 16 } else { 8 };
    let mut offset = moov.0 + header_len;
    let limit = moov.1.min(data.len());

    while offset + 8 <= limit {
        let (size, hlen) = atom_size(data, offset);
        if size < hlen as u64 {
            break;
        }
        let atom_type: &[u8; 4] = data[offset + 4..offset + 8].try_into().unwrap();
        let end = offset + size as usize;

        if atom_type == b"trak" {
            let trak_range = (offset, end.min(limit));
            if is_video_trak(data, trak_range) {
                return Some(trak_range);
            }
        }

        if end > limit {
            break;
        }
        offset = end;
    }
    None
}
